anchor-tx/src/cas_remote.rs

```diff
@@ -0,0 +1,313 @@
+    stream_id: StreamId,
+    cid: String,
+    timestamp: String,
+    ceramic_one_version: String,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Header {
+    pub alg: String,
+    pub kid: String,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Claims {
+    nonce: String,
+    url: String,
+    digest: String,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(rename_all = "camelCase")]
+struct AnchorResponse {
+    pub message: String,
+    pub witness_car: Option<String>,
+}
+
+fn cid_to_stream_id(cid: Cid) -> StreamId {
+    StreamId {
+        r#type: StreamIdType::Unloadable,
+        cid,
+    }
+}
+
+async fn auth_jwt(
+    request_url: String,
+    digest: String,
+    signing_key: &Ed25519KeyPair,
+) -> Result<String> {
+    let controller = did_key_from_ed25519_key_pair(signing_key);
+    let header = Header {
+        kid: format!(
+            "{}#{}",
+            controller,
+            controller
+                .strip_prefix("did:key:")
+                .context("invalid did:key")?
+        ),
+        alg: "EdDSA".to_string(),
+    };
+    let body = Claims {
+        digest,
+        nonce: Uuid::new_v4().to_string(),
+        url: request_url,
+    };
+
+    let header_b64 = b64.encode(serde_json::to_vec(&header)?);
+    let body_b64 = b64.encode(serde_json::to_vec(&body)?);
+    let message = [header_b64, body_b64].join(".");
+    let sig_bytes = signing_key.sign(message.as_bytes());
+    let sig_b64 = b64.encode(sig_bytes);
+    Ok([message.clone(), sig_b64].join("."))
+}
+
+/// Remote CAS transaction manager
+pub struct RemoteCas {
+    signing_key: Ed25519KeyPair,
+    cas_api_url: String,
+}
+
+enum CasResponseParseResult {
+    Anchored(Box<Receipt>),
+    Unauthorized,
+}
+
+#[async_trait]
+impl TransactionManager for RemoteCas {
+    async fn make_proof(&self, root: Cid) -> Result<Receipt> {
+        let loop_count = 12;
+        let wait_duration_min = 5;
+        let mut interval = interval(Duration::from_secs(wait_duration_min * 60));
+        for _ in 0..loop_count {
+            let anchor_response = self.create_anchor_request(root).await?;
+            match parse_anchor_response(anchor_response).await {
+                Ok(CasResponseParseResult::Anchored(receipt)) => {
+                    return Ok(*receipt);
+                }
+                Ok(CasResponseParseResult::Unauthorized) => {
+                    return Err(anyhow!("remote CAS request unauthorized"));
+                }
+                Err(e) => {
+                    debug!("swallowing anchoring result: {}", e);
+                }
+            }
+            interval.tick().await;
+        }
+        Err(anyhow::anyhow!(
+            "{} not anchored after {} {} minute attempts",
+            root.to_string(),
+            loop_count,
+            wait_duration_min
+        ))
+    }
+}
+
+impl RemoteCas {
+    /// Create a new RemoteCas instance
+    pub fn new(keypair: Ed25519KeyPair, cas_api_url: String) -> Self {
```

Comment:
would using a `DidDocument` here instead of a key make sense? it feels like we have a lot of variations on the crypto primitives already - ssi uses ed25519_dalek (and probably ring somewhere) and I used that one in the validation library (continuing this comment/thought on the key.rs file)